cpp/src/parquet/encoding.cc

```diff
@@ -1335,13 +1420,7 @@ class PlainByteArrayDecoder : public PlainDecoder<ByteArrayType>,
           if (ARROW_PREDICT_FALSE(len_ < increment)) {
             ParquetException::EofException();
           }
-          if (ARROW_PREDICT_FALSE(!helper.CanFit(value_len))) {
-            // This element would exceed the capacity of a chunk
-            RETURN_NOT_OK(helper.PushChunk());
-            RETURN_NOT_OK(helper.builder->Reserve(num_values - i));
-            RETURN_NOT_OK(helper.builder->ReserveData(
-                std::min<int64_t>(len_, helper.chunk_space_remaining)));
```

Comment:
Why previously here `ReserveData` for `min(len_, helper.chunk_space_remaining)` here, wouldn't `len_` be too large @pitrou

Reply:
I don't understand: are you commented on the removed code? I'd rather not try to understand code that was removed months ago...

Reply:
See https://github.com/apache/arrow/pull/38437 for context

Reply:
I also don't understand them T_T.

Just confused by this change, so I tried to understand the origin code and find out why this cause the regression, what should I do to fix it